packages/core/test/forbidden.test.ts

```diff
@@ -0,0 +1,77 @@
+import { Machine, assign } from '../src';
+
+interface SignInContext {
+  email: string;
+  password: string;
+}
+
+type ChangePassword = {
+  type: 'changePassword';
+  password: string;
+};
+
+// Needed to put this in a variable to not get the "duplicated property" eslint
+// warning
+const changePassword = 'changePassword';
+
+const authMachine = Machine<SignInContext>(
+  {
+    context: { email: '', password: '' },
+    initial: 'passwordField',
+    states: {
+      passwordField: {
+        initial: 'hidden',
+        states: {
+          hidden: {
+            on: {
+              // We don't want the password field to be revealed based on a
+              // browser autocomplete firing an input `onChange` event.
+              [changePassword]: undefined,
+              // However, we do want to assign the value. Just prevent the
+              // parent transitions
+              [changePassword]: {
+                actions: 'assignPassword'
+              }
```

Comment:
If we were to move the forbidden event underneath the assign action, `assignPassword` will never be called.

The ordering playing a role here makes me think that this is indeed an accidental feature and not intended. 😁

Reply:
The ordering play role here, because `{foo:1,foo:42}` is really `{foo:42}`. You can't have object with multiple properties with the same name after all.

I think the solution for your problem is just this:
```js
{
    [changePassword]: {
        // We don't want the password field to be revealed based on a
        // browser autocomplete firing an input `onChange` event.
        target: undefined,
        // However, we do want to assign the value. Just prevent the
        // parent transitions
        actions: 'assignPassword'
    },
}
```

Reply:
Ok, I've actually just re-read your issue carefully and I think I understand the confusion now. It indeed looks a little bit confusing at first and I would stick to the explicit version that I have written above. It's just not possible for multiple transitions to be taken from different nodes in the ancestor-descendant chain - only the most inner matched transition is taken. And you have matched this transition `[changePassword]: {actions: 'assignPassword'}` which doesn't have a target specified, so it's basically equivalent to what I have written above.

Reply:
Thank you for the reply and clarification. For some reason, it didn't dawn on me that forbidden events can be achieved with `target: undefined` while still defining actions.
I must've been too caught up in my code-style of never using the `[event]: target` syntax that I couldn't make this connection. 😅

I think part of my confusion also came from the docs that say `forbidden_event: undefined` is equivalent to `actions: []`. I'll see if I can change the docs with my new learnings in the PR later.

Should I change my test to your snippet and include it? Might be useful to have a test like this as I couldn't see an existing one testing similar behavior.

If you want to include the test, should I also move it to a different file? Which test suite works best here? I'm thinking `actions` or `event` might work best.

Reply:
> I think part of my confusion also came from the docs that say forbidden_event: undefined is equivalent to actions: []. I'll see if I can change the docs with my new learnings in the PR later.

It's kinda true as all of:
- `forbidden_event: undefined`
- `forbidden_event: ''` (I think it also works, but I have not been using it)
- `forbidden_event: { actions: [] }`
- `forbidden_event: { target: undefined }`
- `forbidden_event: { target: undefined, actions: [] }`

are the same thing. After all empty actions array is like no actions at all.

> Should I change my test to your snippet and include it? Might be useful to have a test like this as I couldn't see an existing one testing similar behavior.

Probably yes, but not sure where it is the best place to put it right now. Try moving it where you think it would be best to put it and we'll iterate from that or just merge in.

Reply:
> It's just not possible for multiple transitions to be taken from different nodes in the ancestor-descendant chain - only the most inner matched transition is taken.

This was the missing and crucial piece for my understanding. I tried to reflect this in the docs and the test too. 👍 

Thank you so much!